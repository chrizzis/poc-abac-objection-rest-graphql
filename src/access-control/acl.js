import { defineAbility } from '@casl/ability'
import Role from './role.js'

/**
 * Rules for authorizing a user requesting to perform an action on a resource.
 * 
 * @param user
 */
// const acl = (user) => { return (resource, body, relation) => { casl shit with baked in user context }}
// also, since casl is coupled to this, should I pull { rulesToQuery, permittedFieldsOf } in here?
function acl(user, resource, body, relation) {
  if (!!resource && !!body) {
    if (resource.modelProperty !== 'acl-user')
      throw new Error('resource is NOT an instance of User class')
    if (relation === 'projects') {
      if (body.modelProperty !== 'acl-project')
        throw new Error('input is NOT an instance of Pet class')
    } else if (body.modelProperty !== 'acl-user')
      throw new Error('input is NOT an instance of User class')
  }

  return defineAbility((allow, forbid) => {
    allow('read', 'User')
    forbid('read', 'User', ['password'])
    allow('read', 'Project', { published: true })
    forbid('read', 'Project', ['ownerOnly'])

    if (user.role === Role.Anonymous) {
      allow('create', 'User')
      // allow('read', 'Project', { published: true })
    } else if (user.role === Role.User) {
      allow('read', 'User', ['password'], { id: user.id })
      allow('update', 'User', { id: user.id })
      allow('delete', 'User', { id: user.id })
      // allow('read', 'Project', { published: true })
      allow('manage', 'Project', { ownerId: user.id })
    } else if (user.role === Role.Admin) {
      allow('read', 'Project')
      allow('manage', 'Project', { ownerId: user.id })
    }

    // TODO: this is unnecessary as it is defined above
    if (relation === 'projects') {
      // resource = user, body = project
      allow('read', 'Project', { id: user.id })
      allow('create', 'Project', { id: user.id })
    } else if (relation === 'owner') {
      // resource = project, body = user
    }
  })
}

export default acl
