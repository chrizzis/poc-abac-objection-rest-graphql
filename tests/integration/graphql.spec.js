import { graphql } from 'graphql'
import createSchema from '../../src/schema.js'
import pool from '../../src/db.js'
import { Model } from 'objection'
import acl from '../../src/access-control/acl.js'

// TODO: something in my graphql implementation is causing yarn test:once to process.exit with 1 after successful testing
// TODO: some async thing? server.close()
// afterFind console.log error: Cannot log after tests are done. Did you forget to wait for something async in your test?
describe('User Access Control', () => {
  let rootNodeForUser = null;
  let runQueryAsUser = null;
  let knex
  let schema
  beforeAll(async () => {
    knex = pool
    Model.knex(pool)
    schema = await createSchema(pool)
  })
  beforeEach(async () => {
    // TODO: this should be an auth middleware for graphql route
    rootNodeForUser = async (username) => {
      let user
      if (username) {
        user = await knex('User').select('id', 'role').where({ username }).first();
      }
      if (!user) {
        user = { role: 'anonymous' }
      }
      // const context = { userId: user.id, isApiQuery: true };
      const context = { _user: user, _isGraphqlQuery: true, _ability: acl(user) } // i think ill have to roll a new fn _graphqlAuth: {authUser, accessibleBy, selectAccessible, whereChain}
      return {
        // TODO: implement this (qb = queryBuilder (AuthQueryBuilder))
        // builder is an objection.js query builder.
        async onQuery(qb) {
          // TODO: objection-graphql uses old-ass versions of objection and graphql.
          // i need to override builder.eager - first i need to trace where it is coming from...
          // qb.eagerAlgorithm(Model.JoinEagerAlgorithm);
          return qb.context(context);
        },
      };
    };
    runQueryAsUser = async (username, query) => {
      const rootNode = await rootNodeForUser(username);
      // Expected undefined to be a GraphQL schema.
      // return graphql(schema, query, rootNode);
      return graphql({ schema, source: query, rootValue: rootNode });
      // return graphql({ schema, source: query });
    };
  });
  it('should have our 3 users in the db', async () => {
    const users = await knex('User').select('username', 'password');
    expect(users).toMatchSnapshot();
  });
  // too simple, deprecate after auth middleware added - entire graphql api should be authed
  it('should not be able to see passwords with an anonymous user', async () => {
    const query = '{users { id, username, password }}'
    // const { data: { users } } = await graphql({ schema, source: query })
    // const { data: { users } } = await runQueryAsUser(null, '{users { id, username, password }}');
    // expect(users).toMatchSnapshot();
    const result = await runQueryAsUser(null, query);
    expect(result).toMatchSnapshot();
  });
  it('should not be able to see the passwords of others with a registered user', async () => {
    // const { data: { users } } = await runQueryAsUser('alice-immutable', '{users { id, username, password }}');
    // expect(users).toMatchSnapshot();
    const result = await runQueryAsUser('alice-immutable', '{users { id, username, password }}');
    expect(result).toMatchSnapshot();
  });
  it('should not be able to see unpublished projects with an anonymous user', async () => {
    // const { data: { projects } } = await runQueryAsUser('alice-immutable', '{projects { id, published, onlyOwner, ownerOnly }}');
    const result = await runQueryAsUser(null, '{projects { id, published, ownerOnly }}');
    expect(result).toMatchSnapshot();
  });
  it('should not be able to see unpublished projects of others with a registered user', async () => {
    // const { data: { projects } } = await runQueryAsUser('alice-immutable', '{projects { id, published, onlyOwner, ownerOnly }}');
    // TODO: had to add ownerId for graphql ABAC plugin to use for acl (how the fuck do i make objection do this on insert?)
    // TODO: i shouldn't have to explicitly request fields just to get the relations -- need to figure out how to fix this...
    const result = await runQueryAsUser('alice-immutable', '{projects { id, published, ownerId, ownerOnly }}');
    // const result = await runQueryAsUser('alice-immutable', '{projects { id, published, ownerId, ownerOnly }}');
    expect(result).toMatchSnapshot();
  });
  it('should not be able to see user passwords, nor unpublished projects with an anonymous user', async () => {
    const query = `{
      users {
        id,
        username,
        password,
        projects {
          title,
          ownerId,
          published,
          ownerOnly
        }
      }
    }`
    const result = await runQueryAsUser(null, query);
    expect(result).toMatchSnapshot();
  });
  it('should not be able to tunnel through graph to see unauthorized data with an anonymous user', async () => {
    const query = `{
      users {
        id,
        username,
        password,
        projects {
          title,
          ownerId,
          published,
          ownerOnly,
          owner {
            id,
            username,
            password,
          }
        }
      }
    }`
    const result = await runQueryAsUser(null, query);
    expect(result).toMatchSnapshot();
  });
})