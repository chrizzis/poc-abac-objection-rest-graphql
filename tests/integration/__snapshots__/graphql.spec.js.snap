// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Access Control should have our 3 users in the db 1`] = `
Array [
  Object {
    "password": "alice",
    "username": "alice-immutable",
  },
  Object {
    "password": "bob",
    "username": "bob-immutable",
  },
  Object {
    "password": "eve",
    "username": "eve-immutable",
  },
]
`;

exports[`User Access Control should not be able to see passwords with an anonymous user 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "id": 1,
        "password": null,
        "username": "alice-immutable",
      },
      Object {
        "id": 2,
        "password": null,
        "username": "bob-immutable",
      },
      Object {
        "id": 3,
        "password": null,
        "username": "eve-immutable",
      },
    ],
  },
}
`;

exports[`User Access Control should not be able to see the passwords of others with a registered user 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "id": 1,
        "password": "alice",
        "username": "alice-immutable",
      },
      Object {
        "id": 2,
        "password": null,
        "username": "bob-immutable",
      },
      Object {
        "id": 3,
        "password": null,
        "username": "eve-immutable",
      },
    ],
  },
}
`;

exports[`User Access Control should not be able to see unpublished projects of others with a registered user 1`] = `
Object {
  "data": Object {
    "projects": Array [
      Object {
        "id": 1,
        "ownerId": 1,
        "ownerOnly": "only alice can see this",
        "published": false,
      },
      Object {
        "id": 4,
        "ownerId": 1,
        "ownerOnly": "only alice can see this",
        "published": true,
      },
      Object {
        "id": 5,
        "ownerId": 2,
        "ownerOnly": null,
        "published": true,
      },
      Object {
        "id": 6,
        "ownerId": 3,
        "ownerOnly": null,
        "published": true,
      },
    ],
  },
}
`;

exports[`User Access Control should not be able to see unpublished projects with an anonymous user 1`] = `
Object {
  "data": Object {
    "projects": Array [
      Object {
        "id": 4,
        "ownerOnly": null,
        "published": true,
      },
      Object {
        "id": 5,
        "ownerOnly": null,
        "published": true,
      },
      Object {
        "id": 6,
        "ownerOnly": null,
        "published": true,
      },
    ],
  },
}
`;

exports[`User Access Control should not be able to see user passwords, nor unpublished projects with an anonymous user 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "id": 1,
        "password": null,
        "projects": Array [
          Object {
            "ownerId": 1,
            "ownerOnly": null,
            "published": true,
            "title": "Published project by Alice",
          },
        ],
        "username": "alice-immutable",
      },
      Object {
        "id": 2,
        "password": null,
        "projects": Array [
          Object {
            "ownerId": 2,
            "ownerOnly": null,
            "published": true,
            "title": "Published project by Bob",
          },
        ],
        "username": "bob-immutable",
      },
      Object {
        "id": 3,
        "password": null,
        "projects": Array [
          Object {
            "ownerId": 3,
            "ownerOnly": null,
            "published": true,
            "title": "Published project by Eve",
          },
        ],
        "username": "eve-immutable",
      },
    ],
  },
}
`;

exports[`User Access Control should not be able to tunnel through graph to see unauthorized data with an anonymous user 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "id": 1,
        "password": null,
        "projects": Array [
          Object {
            "owner": Object {
              "id": 1,
              "password": null,
              "username": "alice-immutable",
            },
            "ownerId": 1,
            "ownerOnly": null,
            "published": true,
            "title": "Published project by Alice",
          },
        ],
        "username": "alice-immutable",
      },
      Object {
        "id": 2,
        "password": null,
        "projects": Array [
          Object {
            "owner": Object {
              "id": 2,
              "password": null,
              "username": "bob-immutable",
            },
            "ownerId": 2,
            "ownerOnly": null,
            "published": true,
            "title": "Published project by Bob",
          },
        ],
        "username": "bob-immutable",
      },
      Object {
        "id": 3,
        "password": null,
        "projects": Array [
          Object {
            "owner": Object {
              "id": 3,
              "password": null,
              "username": "eve-immutable",
            },
            "ownerId": 3,
            "ownerOnly": null,
            "published": true,
            "title": "Published project by Eve",
          },
        ],
        "username": "eve-immutable",
      },
    ],
  },
}
`;
